### ISSUE TITLE
java.lang.IllegalStateException: close() has been called on the Graph

### ISSUE THREAD

**Issue by frauzufall** [imagej/imagej-tensorflow#23]
The recent fix to actually add models to the cache map (https://github.com/imagej/imagej-tensorflow/commit/14f6da185d8c06aac97934c3f2079d8503b98b2f) leads to this potential issue: A `SavedModelBundle` has to be closed to release memory allocation. If an implementation does this:
```
SavedModelBundle model = tensorFlowService.loadModel(source, modelName, MODEL_TAG);
model.close();
model = tensorFlowService.loadModel(source, modelName, MODEL_TAG);
```

.. the `DefaultTensorFlowService` will not know that this model was already closed and return it from the cache map, leading to this exception:
```
java.lang.IllegalStateException: close() has been called on the Graph
```

The solution is not to close the model and let the `DefaultTensorFlowService` close it on `dispose`. One can also trigger closing all cached models by calling `tensorFlowService.dispose()`. 

I thought writing this down in an issue might help someone stumbling over this exception. Is this enough? Also, this is a breaking API change, right..? An alternative would be to not cache models at all. Sadly there seems to be no way to find out if a model was closed. 

Ping @HedgehogCode @ctrueden

**Comment by ctrueden** [comment]
@frauzufall The `SavedModelBundle` is not a final class, so we can create a subclass `ImageJSavedModelBundle extends SavedModelBundle` and override `close()` to invalidate itself from the cache. And then make sure the `TensorFlowService` always wraps models it loads.

**Comment by frauzufall** [comment]
@ctrueden thanks, I tried it but the `SavedModelBundle` constructor is private.. Any way around that?

**Comment by frauzufall** [comment]
An alternative would be some kind of reflection hack to make `SavedModelBundle` usable. Is that better than releasing `2.0.0`?

**Comment by ctrueden** [comment]
Hmm, too bad. OK, so then we can use composition rather than inheritance. Make our own model class that also implements `AutoCloseable` and does the right cleanup when `close()` is called, and has a method `SavedModelBundle model()` to access the TensorFlow `SavedModelBundle`. When our model class's `close()` method is called, it also calls `close()` on the wrapped `SavedModelBundle`. Only pass around our model objects with our API.

**Comment by frauzufall** [comment]
Thanks for the feedback @ctrueden, I made the suggested changes which also restore backwards compatibility by introducing a new `loadCachedModel` method and deprecating `loadModel`, see PR #24

**Comment by ctrueden** [comment]
I merged #24, so closing this.
